File: src/Parsimonie.Api/Services/Roster/RosterService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Parsimonie.Api.Infrastructure.Data;
using Parsimonie.Api.Models.DTOs.Roster;
using Parsimonie.Api.Models.Entities;
using Parsimonie.Api.Models.Enums;
using Parsimonie.Api.Services.Roster.Interfaces;

namespace Parsimonie.Api.Services.Roster;

/// <summary>
/// Service implementation for roster/character management operations
/// </summary>
public class RosterService : IRosterService
{
    private readonly ParsimonieDbContext _dbContext;
    private readonly ILogger<RosterService> _logger;

    public RosterService(ParsimonieDbContext dbContext, ILogger<RosterService> logger)
    {
        _dbContext = dbContext;
        _logger = logger;
    }

    public async Task<RosterResponse> GetRosterAsync(bool includeInactive, bool isOfficer)
    {
        var query = _dbContext.Characters
            .Include(c => c.User)
            .AsQueryable();
        
        // Non-officers only see active characters
        if (!isOfficer || !includeInactive)
        {
            query = query.Where(c => c.IsActive);
        }

        var characters = await query
            .OrderBy(c => c.Class)
            .ThenByDescending(c => c.IsMain)
            .ThenBy(c => c.Name)
            .Select(c => MapToResponse(c, isOfficer))
            .ToListAsync();

        var totalCount = await _dbContext.Characters.CountAsync();
        var activeCount = await _dbContext.Characters.Where(c => c.IsActive).CountAsync();

        return new RosterResponse(characters, totalCount, activeCount);
    }

    public async Task<ServiceResult<CharacterResponse>> GetCharacterAsync(Guid id, bool isOfficer)
    {
        var character = await _dbContext.Characters
            .Include(c => c.User)
            .FirstOrDefaultAsync(c => c.Id == id);

        if (character == null)
        {
            return ServiceResult<CharacterResponse>.NotFound("Character not found");
        }

        // Non-officers can't see inactive characters
        if (!character.IsActive && !isOfficer)
        {
            return ServiceResult<CharacterResponse>.NotFound("Character not found");
        }

        return ServiceResult<CharacterResponse>.Ok(MapToResponse(character, isOfficer));
    }

    public async Task<ServiceResult<CharacterResponse>> CreateCharacterAsync(CreateCharacterRequest request, Guid createdByUserId)
    {
        // Validate name
        if (string.IsNullOrWhiteSpace(request.Name) || request.Name.Length < 2 || request.Name.Length > 12)
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_name", "Character name must be 2-12 characters");
        }

        // Check for duplicate - using ToUpperInvariant for EF Core SQL translation (case-insensitive)
        var nameLower = request.Name.ToUpperInvariant();
        var realmLower = request.Realm.ToUpperInvariant();
#pragma warning disable CA1862 // EF Core database query - StringComparison cannot be translated to SQL
        var exists = await _dbContext.Characters
            .AnyAsync(c => c.Name.ToUpperInvariant() == nameLower && c.Realm.ToUpperInvariant() == realmLower);
#pragma warning restore CA1862
        
        if (exists)
        {
            return ServiceResult<CharacterResponse>.Fail("character_exists", "Character already exists on this realm");
        }

        // Validate spec matches class
        if (!IsValidSpecForClass(request.Class, request.PrimarySpec))
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_spec", "Primary spec is not valid for this class");
        }

        if (request.SecondarySpec.HasValue && !IsValidSpecForClass(request.Class, request.SecondarySpec.Value))
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_spec", "Secondary spec is not valid for this class");
        }

        // If setting as main and user is specified, unset other mains
        if (request.IsMain && request.UserId.HasValue)
        {
            await ClearMainForUserAsync(request.UserId.Value);
        }

        var character = new Character
        {
            Id = Guid.NewGuid(),
            UserId = request.UserId,
            Name = request.Name,
            Realm = request.Realm,
            Class = request.Class,
            PrimarySpec = request.PrimarySpec,
            SecondarySpec = request.SecondarySpec,
            IsMain = request.IsMain,
            IsActive = true,
            Notes = request.Notes,
            CreatedByUserId = createdByUserId,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _dbContext.Characters.Add(character);
        await _dbContext.SaveChangesAsync();

        _logger.LogInformation("Character {CharName}-{Realm} created by {OfficerId}", 
            character.Name, character.Realm, createdByUserId);

        // Reload with user for response
        await _dbContext.Entry(character).Reference(c => c.User).LoadAsync();

        return ServiceResult<CharacterResponse>.Ok(MapToResponse(character, true));
    }

    public async Task<ServiceResult<CharacterResponse>> UpdateCharacterAsync(Guid id, UpdateCharacterRequest request)
    {
        var character = await _dbContext.Characters
            .Include(c => c.User)
            .FirstOrDefaultAsync(c => c.Id == id);

        if (character == null)
        {
            return ServiceResult<CharacterResponse>.NotFound("Character not found");
        }

        // Check for duplicate name/realm if changing
        if (!string.IsNullOrEmpty(request.Name) && 
            (!string.Equals(request.Name, character.Name, StringComparison.OrdinalIgnoreCase) || 
             !string.Equals(request.Realm, character.Realm, StringComparison.OrdinalIgnoreCase)))
        {
            var nameLower = request.Name.ToUpperInvariant();
            var realmLower = request.Realm.ToUpperInvariant();
#pragma warning disable CA1862 // EF Core database query - StringComparison cannot be translated to SQL
            var exists = await _dbContext.Characters
                .AnyAsync(c => c.Id != id && c.Name.ToUpperInvariant() == nameLower && c.Realm.ToUpperInvariant() == realmLower);
#pragma warning restore CA1862
            
            if (exists)
            {
                return ServiceResult<CharacterResponse>.Fail("character_exists", "Character already exists on this realm");
            }
            
            character.Name = request.Name;
            character.Realm = request.Realm;
        }

        // Validate specs
        if (!IsValidSpecForClass(character.Class, request.PrimarySpec))
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_spec", "Primary spec is not valid for this class");
        }

        if (request.SecondarySpec.HasValue && !IsValidSpecForClass(character.Class, request.SecondarySpec.Value))
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_spec", "Secondary spec is not valid for this class");
        }

        // If setting as main, unset other mains
        if (request.IsMain && !character.IsMain && character.UserId.HasValue)
        {
            await ClearMainForUserAsync(character.UserId.Value, character.Id);
        }

        character.PrimarySpec = request.PrimarySpec;
        character.SecondarySpec = request.SecondarySpec;
        character.IsMain = request.IsMain;
        character.IsActive = request.IsActive;
        character.Notes = request.Notes;
        character.UpdatedAt = DateTime.UtcNow;

        await _dbContext.SaveChangesAsync();

        return ServiceResult<CharacterResponse>.Ok(MapToResponse(character, true));
    }

    public async Task<ServiceResult> DeleteCharacterAsync(Guid id, Guid deletedByUserId)
    {
        var character = await _dbContext.Characters.FindAsync(id);
        
        if (character == null)
        {
            return ServiceResult.Fail("character_not_found", "Character not found");
        }

        character.IsActive = false;
        character.UpdatedAt = DateTime.UtcNow;
        await _dbContext.SaveChangesAsync();

        _logger.LogInformation("Character {CharId} soft-deleted by {OfficerId}", id, deletedByUserId);

        return ServiceResult.Ok();
    }

    public async Task<ServiceResult<CharacterResponse>> AssignCharacterAsync(Guid id, Guid? userId, Guid assignedByUserId)
    {
        var character = await _dbContext.Characters
            .Include(c => c.User)
            .FirstOrDefaultAsync(c => c.Id == id);

        if (character == null)
        {
            return ServiceResult<CharacterResponse>.NotFound("Character not found");
        }

        // Validate user exists if assigning
        if (userId.HasValue)
        {
            var userExists = await _dbContext.Users.AnyAsync(u => u.Id == userId.Value);
            if (!userExists)
            {
                return ServiceResult<CharacterResponse>.Fail("user_not_found", "User not found");
            }

            // If character is main, clear other mains for this user
            if (character.IsMain)
            {
                await ClearMainForUserAsync(userId.Value, character.Id);
            }
        }

        character.UserId = userId;
        character.UpdatedAt = DateTime.UtcNow;
        await _dbContext.SaveChangesAsync();

        // Reload user for response
        await _dbContext.Entry(character).Reference(c => c.User).LoadAsync();

        _logger.LogInformation("Character {CharId} assigned to user {UserId} by {OfficerId}", 
            id, userId, assignedByUserId);

        return ServiceResult<CharacterResponse>.Ok(MapToResponse(character, true));
    }

    public async Task<ServiceResult<CharacterResponse>> UpdateSpecAsync(Guid characterId, WowSpec primarySpec, WowSpec? secondarySpec, Guid userId)
    {
        var character = await _dbContext.Characters
            .Include(c => c.User)
            .FirstOrDefaultAsync(c => c.Id == characterId);

        if (character == null || !character.IsActive)
        {
            return ServiceResult<CharacterResponse>.NotFound("Character not found");
        }

        // Check ownership
        if (character.UserId != userId)
        {
            return ServiceResult<CharacterResponse>.Forbidden("You can only update your own characters");
        }

        // Validate specs
        if (!IsValidSpecForClass(character.Class, primarySpec))
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_spec", "Primary spec is not valid for this class");
        }

        if (secondarySpec.HasValue && !IsValidSpecForClass(character.Class, secondarySpec.Value))
        {
            return ServiceResult<CharacterResponse>.Fail("invalid_spec", "Secondary spec is not valid for this class");
        }

        character.PrimarySpec = primarySpec;
        character.SecondarySpec = secondarySpec;
        character.UpdatedAt = DateTime.UtcNow;

        await _dbContext.SaveChangesAsync();

        return ServiceResult<CharacterResponse>.Ok(MapToResponse(character, false));
    }

    public async Task<ServiceResult<CharacterResponse>> SetAsMainAsync(Guid characterId, Guid userId)
    {
        var character = await _dbContext.Characters
            .Include(c => c.User)
            .FirstOrDefaultAsync(c => c.Id == characterId);

        if (character == null || !character.IsActive)
        {
            return ServiceResult<CharacterResponse>.NotFound("Character not found");
        }

        // Check ownership
        if (character.UserId != userId)
        {
            return ServiceResult<CharacterResponse>.Forbidden("You can only modify your own characters");
        }

        // Clear other mains
        await ClearMainForUserAsync(userId, character.Id);

        character.IsMain = true;
        character.UpdatedAt = DateTime.UtcNow;

        await _dbContext.SaveChangesAsync();

        return ServiceResult<CharacterResponse>.Ok(MapToResponse(character, false));
    }

    #region Private Helpers

    private async Task ClearMainForUserAsync(Guid userId, Guid? excludeCharacterId = null)
    {
        var mains = await _dbContext.Characters
            .Where(c => c.UserId == userId && c.IsMain && c.Id != excludeCharacterId)
            .ToListAsync();

        foreach (var main in mains)
        {
            main.IsMain = false;
            main.UpdatedAt = DateTime.UtcNow;
        }
    }

    private static CharacterResponse MapToResponse(Character c, bool includeNotes)
    {
        return new CharacterResponse(
            c.Id,
            c.Name,
            c.Realm,
            c.Class,
            GetClassName(c.Class),
            c.PrimarySpec,
            GetSpecName(c.PrimarySpec),
            c.SecondarySpec,
            c.SecondarySpec.HasValue ? GetSpecName(c.SecondarySpec.Value) : null,
            c.IsMain,
            c.IsActive,
            includeNotes ? c.Notes : null,
            c.User != null ? new CharacterOwnerResponse(c.User.Id, c.User.DiscordUsername, c.User.DiscordAvatar) : null
        );
    }

    private static bool IsValidSpecForClass(WowClass wowClass, WowSpec spec)
    {
        var specValue = (int)spec;
        var classValue = (int)wowClass;
        
        // Specs are numbered as ClassId * 100 + SpecIndex
        return specValue / 100 == classValue;
    }

    private static string GetClassName(WowClass wowClass) => wowClass switch
    {
        WowClass.Warrior => "Warrior",
        WowClass.Paladin => "Paladin",
        WowClass.Hunter => "Hunter",
        WowClass.Rogue => "Rogue",
        WowClass.Priest => "Priest",
        WowClass.Shaman => "Shaman",
        WowClass.Mage => "Mage",
        WowClass.Warlock => "Warlock",
        WowClass.Druid => "Druid",
        _ => "Unknown"
    };

    private static string GetSpecName(WowSpec spec) => spec switch
    {
        // Warriors
        WowSpec.Arms => "Arms",
        WowSpec.Fury => "Fury",
        WowSpec.ProtectionWarrior => "Protection",
        // Paladins
        WowSpec.HolyPaladin => "Holy",
        WowSpec.ProtectionPaladin => "Protection",
        WowSpec.Retribution => "Retribution",
        // Hunters
        WowSpec.BeastMastery => "Beast Mastery",
        WowSpec.Marksmanship => "Marksmanship",
        WowSpec.Survival => "Survival",
        // Rogues
        WowSpec.Assassination => "Assassination",
        WowSpec.Combat => "Combat",
        WowSpec.Subtlety => "Subtlety",
        // Priests
        WowSpec.Discipline => "Discipline",
        WowSpec.HolyPriest => "Holy",
        WowSpec.Shadow => "Shadow",
        // Shamans
        WowSpec.Elemental => "Elemental",
        WowSpec.Enhancement => "Enhancement",
        WowSpec.RestorationShaman => "Restoration",
        // Mages
        WowSpec.Arcane => "Arcane",
        WowSpec.Fire => "Fire",
        WowSpec.FrostMage => "Frost",
        // Warlocks
        WowSpec.Affliction => "Affliction",
        WowSpec.Demonology => "Demonology",
        WowSpec.Destruction => "Destruction",
        // Druids
        WowSpec.Balance => "Balance",
        WowSpec.FeralCombat => "Feral",
        WowSpec.RestorationDruid => "Restoration",
        _ => "Unknown"
    };

    #endregion
}

File: src/Parsimonie.Web/src/app/core/services/roster.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { WowClass, WowSpec } from '../models/user.model';

export interface CharacterOwner {
  id: string;
  username: string;
  avatar: string | null;
}

export interface CharacterResponse {
  id: string;
  name: string;
  realm: string;
  class: WowClass;
  className: string;
  primarySpec: WowSpec;
  primarySpecName: string;
  secondarySpec: WowSpec | null;
  secondarySpecName: string | null;
  isMain: boolean;
  isActive: boolean;
  notes: string | null;
  owner: CharacterOwner | null;
}

export interface RosterResponse {
  characters: CharacterResponse[];
  totalCount: number;
  activeCount: number;
}

export interface CreateCharacterRequest {
  name: string;
  realm: string;
  class: WowClass;
  primarySpec: WowSpec;
  secondarySpec?: WowSpec;
  userId?: string;
  isMain: boolean;
  notes?: string;
}

export interface UpdateCharacterRequest {
  name?: string;
  realm: string;
  primarySpec: WowSpec;
  secondarySpec?: WowSpec;
  isMain: boolean;
  isActive: boolean;
  notes?: string;
}

export interface UpdateSpecRequest {
  primarySpec: WowSpec;
  secondarySpec?: WowSpec;
}

export interface AssignCharacterRequest {
  userId: string | null;
}

@Injectable({
  providedIn: 'root'
})
export class RosterService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/roster`;

  // Get all roster characters
  getRoster(includeInactive = false): Observable<RosterResponse> {
    return this.http.get<RosterResponse>(`${this.baseUrl}?includeInactive=${includeInactive}`);
  }

  // Get single character
  getCharacter(id: string): Observable<CharacterResponse> {
    return this.http.get<CharacterResponse>(`${this.baseUrl}/${id}`);
  }

  // Create character (Officer only)
  createCharacter(request: CreateCharacterRequest): Observable<CharacterResponse> {
    return this.http.post<CharacterResponse>(this.baseUrl, request);
  }

  // Update character (Officer only)
  updateCharacter(id: string, request: UpdateCharacterRequest): Observable<CharacterResponse> {
    return this.http.put<CharacterResponse>(`${this.baseUrl}/${id}`, request);
  }

  // Delete character (Officer only - soft delete)
  deleteCharacter(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // Assign character to user (Officer only)
  assignCharacter(id: string, userId: string | null): Observable<CharacterResponse> {
    return this.http.put<CharacterResponse>(`${this.baseUrl}/${id}/assign`, { userId });
  }

  // Update own character spec (Raider)
  updateSpec(id: string, request: UpdateSpecRequest): Observable<CharacterResponse> {
    return this.http.put<CharacterResponse>(`${this.baseUrl}/${id}/spec`, request);
  }

  // Set character as main (Raider)
  setAsMain(id: string): Observable<CharacterResponse> {
    return this.http.put<CharacterResponse>(`${this.baseUrl}/${id}/main`, {});
  }
}

// Helper functions
export const WOW_CLASS_NAMES: Record<WowClass, string> = {
  [WowClass.Warrior]: 'Warrior',
  [WowClass.Paladin]: 'Paladin',
  [WowClass.Hunter]: 'Hunter',
  [WowClass.Rogue]: 'Rogue',
  [WowClass.Priest]: 'Priest',
  [WowClass.Shaman]: 'Shaman',
  [WowClass.Mage]: 'Mage',
  [WowClass.Warlock]: 'Warlock',
  [WowClass.Druid]: 'Druid',
};

export const WOW_CLASS_COLORS: Record<WowClass, string> = {
  [WowClass.Warrior]: '#C79C6E',
  [WowClass.Paladin]: '#F58CBA',
  [WowClass.Hunter]: '#ABD473',
  [WowClass.Rogue]: '#FFF569',
  [WowClass.Priest]: '#FFFFFF',
  [WowClass.Shaman]: '#0070DE',
  [WowClass.Mage]: '#69CCF0',
  [WowClass.Warlock]: '#9482C9',
  [WowClass.Druid]: '#FF7D0A',
};

export const WOW_SPECS_BY_CLASS: Record<WowClass, { value: WowSpec; name: string }[]> = {
  [WowClass.Warrior]: [
    { value: WowSpec.Arms, name: 'Arms' },
    { value: WowSpec.Fury, name: 'Fury' },
    { value: WowSpec.ProtectionWarrior, name: 'Protection' },
  ],
  [WowClass.Paladin]: [
    { value: WowSpec.HolyPaladin, name: 'Holy' },
    { value: WowSpec.ProtectionPaladin, name: 'Protection' },
    { value: WowSpec.Retribution, name: 'Retribution' },
  ],
  [WowClass.Hunter]: [
    { value: WowSpec.BeastMastery, name: 'Beast Mastery' },
    { value: WowSpec.Marksmanship, name: 'Marksmanship' },
    { value: WowSpec.Survival, name: 'Survival' },
  ],
  [WowClass.Rogue]: [
    { value: WowSpec.Assassination, name: 'Assassination' },
    { value: WowSpec.Combat, name: 'Combat' },
    { value: WowSpec.Subtlety, name: 'Subtlety' },
  ],
  [WowClass.Priest]: [
    { value: WowSpec.Discipline, name: 'Discipline' },
    { value: WowSpec.HolyPriest, name: 'Holy' },
    { value: WowSpec.Shadow, name: 'Shadow' },
  ],
  [WowClass.Shaman]: [
    { value: WowSpec.Elemental, name: 'Elemental' },
    { value: WowSpec.Enhancement, name: 'Enhancement' },
    { value: WowSpec.RestorationShaman, name: 'Restoration' },
  ],
  [WowClass.Mage]: [
    { value: WowSpec.Arcane, name: 'Arcane' },
    { value: WowSpec.Fire, name: 'Fire' },
    { value: WowSpec.FrostMage, name: 'Frost' },
  ],
  [WowClass.Warlock]: [
    { value: WowSpec.Affliction, name: 'Affliction' },
    { value: WowSpec.Demonology, name: 'Demonology' },
    { value: WowSpec.Destruction, name: 'Destruction' },
  ],
  [WowClass.Druid]: [
    { value: WowSpec.Balance, name: 'Balance' },
    { value: WowSpec.FeralCombat, name: 'Feral' },
    { value: WowSpec.RestorationDruid, name: 'Restoration' },
  ],
};

export const ALL_WOW_CLASSES = [
  WowClass.Warrior,
  WowClass.Paladin,
  WowClass.Hunter,
  WowClass.Rogue,
  WowClass.Priest,
  WowClass.Shaman,
  WowClass.Mage,
  WowClass.Warlock,
  WowClass.Druid,
];
